dhparams module: initialize

Ekleog opened the pull request:
###### Motivation for this change

Generating per-computer Diffie-Hellman parameters is both something advisable in security, and something that is slow and painful to do by hand.

This also fixes https://github.com/NixOS/nixpkgs/issues/22632

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

(Most checkboxes just don't apply to this module-only PR)

Hope this helps!
Leo

Ekleog committed:
dhparams module: initialize

joachifm commented:
See also https://github.com/NixOS/nixpkgs/pull/12912

Ekleog commented:
OK, I have just been told about https://github.com/NixOS/nixpkgs/pull/11505 and https://github.com/NixOS/nixpkgs/pull/12912 . Compared to the first one, it avoids non-determinism, and compared to the second one, it is way shorter.

The aim is that it makes services using this as a backend (like nginx could, by default) avoid having to bother about cleaning up the state when the dhparams are no longer needed.

3noch commented:
In the absence of nginx doing this by itself, I would use this. (I.e. I need this for nginx.)

joachifm commented on nixos/modules/security/dhparams.nix:
```diff
@@ -0,0 +1,56 @@
+          Diffie-Hellman parameters to generate.
+
+          The value is the size (in bits) of the DH params to generate. The
+          generated DH params path can be found in `${dir}/[name].pem`
```
If you want markup, please use the appropriate docbook elements here, `<filename>` perhaps.

joachifm replied:
Something like `<filename>${dir}/<replaceable>name</replaceable>.pem</filename>` might do it.

3noch replied:
Could `dir` be configurable or at least exposed somehow? I'd rather not hard-code this path but use something like `${security.dhparams.name.path}` in my nginx config.

Ekleog replied:
Fixed, thanks!

Ekleog replied:
Hmm, I get the issue but can't think of a clean UI to do it. Adding a `.path` requires to move the size to a .size argument (that's OK), but then people could override the `.path` argument… which would not be an issue, if I had any idea of how to make a default value for an argument dependent on the name of its parent (to be able to default to `${dir}/[name].pem`.

Do you have an idea? Or would just putting a `security.dhparamsPath` defaulting to `/var/cache/dhparams` be OK with your use case?

abbradar commented:
If we want to have a generic solution like my old PR I think that doing that in systemd is better since it greatly improves boot/switch times (imagine you'd need to wait for 3-4 dhparams to be created if done via activation script vs. doing this concurrently with other dhparams and other services).

3noch commented:
@Ekleog I'd be willing to help push these changes through. Let me know.

Ekleog commented:
@abbradar: That's completely true, your old PR cut down to the bare minimum (DH params don't actually need to be hidden from anyone in any threat model that I know of, and the main issue with your PR was that it was too long, as far as I read it) is most likely a better option than this one. I can also rewrite this PR as a couple of systemd services, as it also handles cleanup of removed dhparams, as you want. (I wouldn't want to "steal" the systemd service idea from you.) How do you feel about taking this forwards?

abbradar commented:
@Ekleog Admittedly I have little time to give love to my old PR nowadays, so I'm completely okay with you stealing the credit ^_^. One small feature that I'd like to be saved is automatic add of `Before` field, e.g. if I create dhparams for Postfix the service would contain `Before=postfix.service`.

cc @edolstra -- how do you feel about this? With old dhparams cleaning this becomes a bit more than my old PR (not a one-line call now :D) I agree that it still may be strange to have this outside nginx, but OTOH a lot of services (any TLS-supporting server in fact) are in need of the same functionality.

Ekleog commented:
OK, I just pushed an overhaul of the PR using a systemd service.

I also added two features I had forgotten about: most importantly, the fact that if the number of bits asked for is changed, the DH parameters have to be rebuilt; and also trying to remove also now-empty directories after removing DH params from the config.

From an UI point of view, I added the `Before=[name].service` as advised, as it is really useful now that it's a systemd service, but I thought that it was not worth adding a dedicated option for overriding this, as it would increase boilerplate both in user code and in the module. Maybe someone would disagree with me on this point, though?

That said, I think the signal/noise ratio of the module is pretty good: it has 11 lines of documentation and 18 lines that would have to be duplicated between all TLS-supporting services for just 18 lines that qualify as non-trivial module-supporting interface (including the 8 lines of non-trivial option definitions).

What do you think about this renewed PR?

fpletz commented:
@Ekleog I think you pushed the wrong commit, I can't see a systemd service in the diff.

I agree with your comments and your basic approach though. :+1:

abbradar commented:
@Ekleog I agree that we don't need a separate option for `Before`. One can use `systemd.services.foo.before` if needed, just mention in the documentation that `Before=[name].service` is already set. It's nice that you support changing bit number!

Ekleog commented:
Oh, I pushed the wrong commit indeed -_- Here it is, thanks for the notice!!

abbradar reviewed (approved):
Haven't tested but looks fine to me. It would be better IMHO if initialization step also happened in a systemd service and others would be `After`/`Wants` to increase parallelism but this way is fine too.

Ekleog commented:
Hmm, is the "approved these changes" a new policy for nixpkgs, is it for making hydra build the PR, or is it for some other reason? It's the first time I've seen this in use.

Also, for putting this in a systemd service, as it's only one call to openssl per dh params that were enabled at last generation, I'm thinking it's negligible in time? (`time sh -c "for i in seq 1 1000; do openssl dhparam -in /tmp/a -text; done"` takes 10ms here, and I can't think of anyone with a use case crazy enough to require 1000 different DH parameters on the same computer)

fpletz commented:
@Ekleog We don't have a policy around the approved changes feature yet.

You may want to use different dhparams per service if you want to make fingerprint TLS harder. For instance if you run multiple tor hidden services you might want to hide the fact that they're running on the same machine. There are other fingerprinting techniques of course, but dhparams is one of them and if you use custom dhparams it's even more important to use different ones. ;)

fpletz reviewed (approved):
Even though I tend to agree with @abbradar in principle I'm also fine with the current solution as an additional activation script.

Ekleog commented:
Here is a version using a systemd service instead of an activation script for old dhparams cleanup. Is this looking all good now? :)

3noch commented:
How's this coming? I'm pretty excited about this.

Ekleog commented:
I don't really know what this is waiting for, but given the 17.03 milestone I'm hopeful, as branch-off is in 3 days :)

fpletz commented:
I was waiting for input from other contributors that didn't open a pull request for a dhparams module before (me and @abbradar basically :smiley:). We're a bit biased on that topic.

Ekleog commented:
Great it's OK, then! 😃 
Thanks for all your feedback!
